Optimize GetPropsFromDict: use tags for conversion

bp-kelley opened the pull request:
This is a test for optimizing GetPropsAsDict, it should be >10x faster in practice if it doesn't segfault the tests.

greglandrum commented on Code/GraphMol/Wrap/props.hpp:
```diff
@@ -76,42 +77,67 @@ bool AddToDict(const U &ob, boost::python::dict &dict, const std::string &key) {
-    }
-    if (AddToDict<std::string>(obj, dict, keys[i])) {
+  for (auto &rdvalue: data) {
+    if (std::find(keys.begin(), keys.end(), rdvalue.key) == keys.end())
```
Out of curiosity: how can it happen that one of the keys from rdvalue is not in `keys`?

greglandrum commented on Code/GraphMol/Wrap/props.hpp:
```diff
@@ -76,42 +77,67 @@ bool AddToDict(const U &ob, boost::python::dict &dict, const std::string &key) {
+      case RDTypeTag::StringTag:
+	if (autoConvert) {
+	  // Auto convert strings to ints and double if possible
+	  if(AddToDict<int>(obj, dict, rdvalue.key)) break;
+	  else if(AddToDict<double>(obj, dict, rdvalue.key)) break;
```
```suggestion
	  if(AddToDict<int>(obj, dict, rdvalue.key)) {
	    break;
	  } else if(AddToDict<double>(obj, dict, rdvalue.key)) { 
	    break;
	  }
```

greglandrum commented on Code/GraphMol/Wrap/props.hpp:
```diff
@@ -76,42 +77,67 @@ bool AddToDict(const U &ob, boost::python::dict &dict, const std::string &key) {
+      case RDTypeTag::VecStringTag:
+	dict[rdvalue.key] = from_rdvalue<std::vector<std::string>>(rdvalue.val);
+	break;
+      }
```
Maybe add a `default` here with with an `UNDER_CONSTRUCTION("not implemented")` invariant to catch the situation where we add a new type tag and forget to update this file?

bp-kelley replied:
Will do.

greglandrum commented on Code/GraphMol/Wrap/Atom.cpp:
```diff
@@ -388,7 +388,8 @@ struct atom_wrapper {
              (python::arg("self"), python::arg("includePrivate") = true,
-              python::arg("includeComputed") = true),
+              python::arg("includeComputed") = true,
+	      python::arg("autoConvert") = true),
```
I think the name of this argument should be `autoConvertStrings` to make it more explicit when the autoconversion actually happens.

greglandrum reviewed (changes requested):
This makes soooo much sense.
small changes requested.

bp-kelley commented:
@greglandrum I was thinking about adding a GetProp(propane, autoConvert=True) or a GetPyProp(propane) in order to get a single value and not having to return the dictionary.

Preferences?

greglandrum commented:
> @greglandrum I was thinking about adding a GetProp(propane, autoConvert=True) or a GetPyProp(propane) in order to get a single value and not having to return the dictionary.
> 
> Preferences?

Nice idea! I like the `autoConvert=True` form better.

greglandrum reviewed (approved):
LGTM

greglandrum commented:
> > @greglandrum I was thinking about adding a GetProp(propane, autoConvert=True) or a GetPyProp(propane) in order to get a single value and not having to return the dictionary.
> > Preferences?
> 
> Nice idea! I like the `autoConvert=True` form better.

I think that's a separate PR though. Or would you prefer to do it in this one?

bp-kelley commented:
Let me see how the code refactors first, otherwise I'll make it a second PR.

bp-kelley commented:
I added the GetProp enhancement, it defaults to False for now.  I added the tests to `testPropertyLists.py` mainly because it pains me to add anything more to rough_test.py :). I had to add C++ test code as well to actually set the properties.  This tests

Mol
Atom
Bond
Conformer

but not substance group.

greglandrum commented:
> I added the tests to `testPropertyLists.py` mainly because it pains me to add anything more to rough_test.py :)

Adding new test files is almost never wrong! :-)

greglandrum reviewed (changes requested):
I think the C++ testing function should be renamed in python, but otherwise this looks good.

greglandrum reviewed (approved):
LGTM
will merge once the CI passes

bp-kelley commented:
@greglandrum it has the same problem I saw before, we do this in other places, any idea why adding the underscore doesn't work here?

bp-kelley commented:
@greglandrum I found the issue, I need to include private vars from their libs not from Chem.  I.e. rdkit.Chem.rdmolops
